core/src/c_api/tiledb.cc

```diff
@@ -117,93 +116,105 @@ int tiledb_ctx_init(
         tiledb_config->write_method_);
 
   // Create storage manager
-  (*tiledb_ctx)->storage_manager_ = new tiledb::StorageManager();
-  tiledb::Status st = (*tiledb_ctx)->storage_manager_->init(config);
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
+  (*ctx)->storage_manager_ = new tiledb::StorageManager();
+  if (save_error(*ctx, (*ctx)->storage_manager_->init(config)))
     return TILEDB_ERR;
-  }
 
   // Success
   return TILEDB_OK;
 }
 
-int tiledb_ctx_finalize(TileDB_CTX* tiledb_ctx) {
+int tiledb_ctx_finalize(TileDB_CTX* ctx) {
   // Trivial case
-  if (tiledb_ctx == nullptr)
+  if (ctx == nullptr)
     return TILEDB_OK;
 
   // Finalize storage manager
   tiledb::Status st;
-  if (tiledb_ctx->storage_manager_ != nullptr)
-    st = tiledb_ctx->storage_manager_->finalize();
+  if (ctx->storage_manager_ != nullptr)
+    st = ctx->storage_manager_->finalize();
+  delete ctx->storage_manager_;
 
-  // Clean up
-  delete tiledb_ctx->storage_manager_;
-  free(tiledb_ctx);
+  if (ctx->last_error_ != nullptr)
+    delete ctx->last_error_;
 
-  // Error
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
-    return TILEDB_ERR;
+  free(ctx);
+
+  return st.ok() ? TILEDB_OK : TILEDB_ERR;
+}
+
+typedef struct tiledb_error_t { const tiledb::Status* status_; } tiledb_error_t;
```

Comment:
Minor comment: I thought we had disabled single-line struct definitions in formatting.